pkg/controller/user-cluster-controller-manager/application-installation-controller/controller.go

```diff
@@ -112,18 +128,116 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
+			return fmt.Errorf("handling deletion of application installation: %w", err)
+		}
+		return nil
+	}
+
+	if err := kuberneteshelper.TryAddFinalizer(ctx, r.userClient, appInstallation, appkubermaticv1.ApplicationInstallationCleanupFinalizer); err != nil {
+		return fmt.Errorf("failed to add finalizer: %w", err)
+	}
+
+	appHasBeenInstalled := appInstallation.Status.ApplicationVersion != nil
+
+	// get applicationDefinition. If it can not be found, there are 2 cases:
+	//   1) KKP admin has removed the applicationDefinition, and we have to remove the corresponding ApplicationInstallation(s)
+	//   2) User made a mistake, or applicationDefinition has not been synced yet on this seed. So we just notify the user.
+	applicationDef := &appkubermaticv1.ApplicationDefinition{}
+	if err := r.seedClient.Get(ctx, types.NamespacedName{Name: appInstallation.Spec.ApplicationRef.Name}, applicationDef); err != nil {
+		if apierrors.IsNotFound(err) {
+			if appHasBeenInstalled {
+				r.traceWarning(appInstallation, log, applicationDefinitionRemovedEvent, fmt.Sprintf("ApplicationDefinition '%s' has been deleted, removing applictionInstallation", applicationDef.Name))
+				return r.userClient.Delete(ctx, appInstallation)
+			} else {
+				return fmt.Errorf("ApplicationDefinition '%s' does not exist. can not install application", applicationDef.Name)
+			}
+		}
+		return err
+	}
+
+	if !applicationDef.DeletionTimestamp.IsZero() {
+		r.traceWarning(appInstallation, log, applicationDefinitionDeletingEvent, fmt.Sprintf("ApplicationDefinition '%s' is being deleted,  removing applictionInstallation", applicationDef.Name))
+		return r.userClient.Delete(ctx, appInstallation)
+	}
+
+	// get applicationVersion. If it can not be found, there are 2 cases:
+	//   1) KKP admin has removed the applicationVersion, and we have to remove the corresponding ApplicationInstallation(s)
+	//   2) User made a mistake, or applicationDefinition has not been synced yet on this seed. So we just notify the user.
+	appVersion := &appkubermaticv1.ApplicationVersion{}
+	if err := r.getApplicationVersion(appInstallation, applicationDef, appVersion); err != nil {
+		if appHasBeenInstalled {
+			r.traceWarning(appInstallation, log, applicationVersionRemovedEvent, fmt.Sprintf("applicationVersion: '%s' has been deleted. removing Application", appInstallation.Spec.ApplicationRef.Version))
+			return r.userClient.Delete(ctx, appInstallation)
+		} else {
+			return fmt.Errorf("applicationVersion: '%s' does not exist. can not install application", appInstallation.Spec.ApplicationRef.Version)
+		}
+	}
+
+	oldAppInstallation := appInstallation.DeepCopy()
+	appInstallation.Status.ApplicationVersion = appVersion
+
+	if err := r.userClient.Status().Patch(ctx, appInstallation, ctrlruntimeclient.MergeFrom(oldAppInstallation)); err != nil {
+		return fmt.Errorf("failed to update status with applicationVersion: %w", err)
+	}
+
+	// install application into the user-cluster
```

Comment:
quick question regarding the comment here: "install" here is meant kind of as an "ensure application is installed", right? Because it could also be that the application is already installed in the correct version and nothing more needs to be done.
I basically am assuming that `handleInstallation` has logic to determine whether any installations need to be done.